### PR TITLE
SOLR-15474 Make Circuit breakers pluggable (take 2)

### PR DESCRIPTION
**Pull request by janhoy**
https://issues.apache.org/jira/browse/SOLR-15474

This is a continuation of the effort from PR #193 which had grown out of date by almost 2 years and no activity.
Thus I updated that branch to main and squashed as one commit to kickstart this PR. I intend to bring it to mergable state.

To do before committable
- [x] Decide whether to move this to a cluster plugin instead of per core
- [x] Re-introduce support for `<circuitBreaker class="solr.CircuitBreakerManager">` with all params  for back compat
- [x] Update ref guide

### CODE REVIEW

File: solr/core/src/java/org/apache/solr/util/circuitbreaker/CPUCircuitBreaker.java
Comment (cpoerschke) [inline solr/core/src/java/org/apache/solr/util/circuitbreaker/CPUCircuitBreaker.java:51]: minor: not 100% sure if this is needed (been doing more non-Java code recently ...)
```suggestion
```

File: solr/core/src/java/org/apache/solr/util/circuitbreaker/CPUCircuitBreaker.java
Comment (cpoerschke) [inline solr/core/src/java/org/apache/solr/util/circuitbreaker/CPUCircuitBreaker.java:54]:
```suggestion
  public void setThreshold(double threshold) {
    this.cpuUsageThreshold = threshold;
```

File: solr/core/src/java/org/apache/solr/util/circuitbreaker/CircuitBreaker.java
Comment (cpoerschke) [inline solr/core/src/java/org/apache/solr/util/circuitbreaker/CircuitBreaker.java:37]:
```suggestion
```

File: solr/core/src/java/org/apache/solr/util/circuitbreaker/CircuitBreakerManager.java
Comment (cpoerschke) [inline solr/core/src/java/org/apache/solr/util/circuitbreaker/CircuitBreakerManager.java:38]: Wondering if the last two paragraphs of the class-level javadoc can be omitted or replaced now then and perhaps the class could even be marked final to make it clear that extending it is no longer intendeds?
```suggestion
public final class CircuitBreakerManager {
```

File: solr/core/src/test-files/solr/collection1/conf/solrconfig-memory-circuitbreaker.xml
Comment (cpoerschke) [inline solr/core/src/test-files/solr/collection1/conf/solrconfig-memory-circuitbreaker.xml:81]:
```suggestion
  <circuitBreaker class="org.apache.solr.util.circuitbreaker.MemoryCircuitBreaker">
```

File: solr/core/src/test-files/solr/collection1/conf/solrconfig-memory-circuitbreaker.xml
Comment (cpoerschke) [inline solr/core/src/test-files/solr/collection1/conf/solrconfig-memory-circuitbreaker.xml:85]:
```suggestion
  <circuitBreaker class="org.apache.solr.util.circuitbreaker.CPUCircuitBreaker">
```

File: solr/core/src/test-files/solr/collection1/conf/solrconfig-memory-circuitbreaker.xml
Comment (cpoerschke) [inline solr/core/src/test-files/solr/collection1/conf/solrconfig-memory-circuitbreaker.xml:86]: Not sure if the `int` might need to be non-`int` here to match the `setThreshold(double threshold)` signature e.g.
```suggestion
    <double  name="threshold">75</double>
```

File: solr/solr-ref-guide/modules/deployment-guide/pages/circuit-breakers.adoc
Comment (cpoerschke) [inline solr/solr-ref-guide/modules/deployment-guide/pages/circuit-breakers.adoc:110]:
```suggestion
<circuitBreaker class="com.mycompany.solr.FooBarCircuitBreaker" enable="true">
```

**Review by cpoerschke — COMMENTED** [review]
Great to see this effort continue!

File: solr/core/src/java/org/apache/solr/util/circuitbreaker/CPUCircuitBreaker.java
Comment (janhoy) [reply solr/core/src/java/org/apache/solr/util/circuitbreaker/CPUCircuitBreaker.java:51]: I wonder if an explicit default constructor is necessary when you do reflection (plugin loading)...

**Comment by janhoy** [comment]
In the last commits I re-introduced backward compatibility for configuring the legacy "CircuitBreakerManager" as a single plugin. I did this by renaming `CircuitBreakerManager` as `CircuitBreakerRegistry`, so the registry is now in charge of knowing about all CBs, while the old Manager is now a true plugin, which wraps the cpu and mem CBs.

I'll probably not pursue converting to cluster-plugin in this PR since it will be a 10.0 only thing. But I want to achieve separating between update and query breakers.

File: solr/core/src/java/org/apache/solr/util/circuitbreaker/CircuitBreakerManager.java
Comment (cpoerschke) [inline solr/core/src/java/org/apache/solr/util/circuitbreaker/CircuitBreakerManager.java:47]: precommit says `/home/runner/work/solr/solr/solr/core/src/java/org/apache/solr/util/circuitbreaker/CircuitBreakerManager.java:47: warning: [OperatorPrecedence] Use grouping parenthesis to make the operator precedence explicit` i.e. presumably that means this:
```suggestion
    return (memEnabled && memCB.isTripped()) || (cpuEnabled && cpuCB.isTripped());
```

**Review by cpoerschke — COMMENTED** [review]
16 / 18 files viewed -- only partly looked at the test changes so far.

And deprecating CircuitBreakerManager via morphing into a wrapper of two circuit breakers is a neat way to maintain backwards compatibility, very nice.

File: solr/core/src/java/org/apache/solr/util/circuitbreaker/MemoryCircuitBreaker.java
Comment (janhoy) [reply solr/core/src/java/org/apache/solr/util/circuitbreaker/MemoryCircuitBreaker.java:50]: As long as an old solrconfig with old-style config works if we switch to double, then it should be ok. I can test.

File: solr/core/src/java/org/apache/solr/util/circuitbreaker/CircuitBreakerRegistry.java
Comment (janhoy) [reply solr/core/src/java/org/apache/solr/util/circuitbreaker/CircuitBreakerRegistry.java:34]: Agree

File: solr/solr-ref-guide/modules/deployment-guide/pages/circuit-breakers.adoc
Comment (janhoy) [reply solr/solr-ref-guide/modules/deployment-guide/pages/circuit-breakers.adoc:39]: Hmm, will have to go a round with this enable/disable config/logic. Probably no need for the toggle at all, if you can simply not add the plugin in the first place.

**Comment by janhoy** [comment]
Thanks for reviewing @cpoerschke. Last bit of polish and then I hope to wrap up. Then this is a nice foundation to build `/update` circuit breaking on top of. I plan to let each CB plugin decide whether it should trip `UPDATE`, `SEARCH` or both request types. So you could e.g. have a CPU breaker on 80% for updates, and 90% for search, letting you sacrifice freshness first and then user experience. But that is for another PR...

**Comment by janhoy** [comment]
Will likely need to squash and force-push in order for Crave.io to do its job... Anyone have a problem with that?

File: solr/solr-ref-guide/modules/deployment-guide/pages/circuit-breakers.adoc
Comment (cpoerschke) [inline solr/solr-ref-guide/modules/deployment-guide/pages/circuit-breakers.adoc:44]:
```suggestion
 <double name="threshold">75</double>
```

File: solr/core/src/test-files/solr/collection1/conf/solrconfig-legacy-circuitbreaker.xml
Comment (janhoy) [reply solr/core/src/test-files/solr/collection1/conf/solrconfig-legacy-circuitbreaker.xml:88]: Sure

**Comment by janhoy** [comment]
@dsmiley, @elyograg: I think this is ready, but appreciate if one of you would skim over this PR and see what you find.

**Comment by elyograg** [comment]
@janhoy this deals with code where I have no knowledge.  I wouldn't know how to tell whether the change is good or not.

**Comment by dsmiley** [comment]
My colleagues have done some custom Circuit Breaking stuff in Solr... I'm hoping one of them might weigh-in here so please don't merge too quickly.

**Comment by dsmiley** [comment]
When you said "Then this is a nice foundation to build /update circuit breaking on top of.", I this caught my interest as this is what we're doing with Solr where I work.  If this PR doesn't _actually_ have to do with that, and is only about configuration, then merge away!

**Comment by janhoy** [comment]
> When you said "Then this is a nice foundation to build /update circuit breaking on top of.", I this caught my interest as this is what we're doing with Solr where I work. If this PR doesn't _actually_ have to do with that, and is only about configuration, then merge away!

Sure, this is just making CBs pluggable for real, so you could create a `DayOfWeekCB` if you want :)

**Comment by janhoy** [comment]
Looking for at least one "+1" before I proceed with the merge.

**Review by atris — APPROVED** [review]

File: solr/core/src/java/org/apache/solr/util/circuitbreaker/CircuitBreakerManager.java
Comment (cpoerschke) [inline solr/core/src/java/org/apache/solr/util/circuitbreaker/CircuitBreakerManager.java:55]:
```suggestion
    if (memEnabled) {
      sb.append(memCB.getDebugInfo());
    }
    if (memEnabled && cpuEnabled) {
      sb.append("\n");
    }
```

File: solr/core/src/test-files/solr/collection1/conf/solrconfig-pluggable-circuitbreaker.xml
Comment (cpoerschke) [inline solr/core/src/test-files/solr/collection1/conf/solrconfig-pluggable-circuitbreaker.xml:82]:
```suggestion
    <double  name="threshold">75</double>
```

File: solr/core/src/test/org/apache/solr/util/BaseTestCircuitBreaker.java
Comment (cpoerschke) [inline solr/core/src/test/org/apache/solr/util/BaseTestCircuitBreaker.java:245]: Could `deregisterAll` be used here instead (and then `getRegisteredCircuitBreakers()` could be removed)?

```suggestion
    h.getCore().getCircuitBreakerRegistry().deregisterAll();
```

**Review by cpoerschke — APPROVED** [review]
LGTM

File: solr/core/src/test/org/apache/solr/util/BaseTestCircuitBreaker.java
Comment (janhoy) [reply solr/core/src/test/org/apache/solr/util/BaseTestCircuitBreaker.java:245]: @cpoerschke Seems we had a race condition with your latest suggestions and my merge. Sorry for that, I could swear I saw some of those comments earlier, don't know how they got closed without merge.

I'll backport the initial commit. Would you mind filing a new PR with these latest changes? The commit message can reference same JIRA.

Comment (cpoerschke) [reply solr/core/src/test/org/apache/solr/util/BaseTestCircuitBreaker.java:245]: No problem! #1870 for the follow-up bits. Would you mind including it in the back port?

Comment (janhoy) [reply solr/core/src/test/org/apache/solr/util/BaseTestCircuitBreaker.java:245]: That was my thought, probably easier to correlate later..

**Pull request merged**
